tsdb/index/postings.go

```diff
@@ -398,38 +398,73 @@ func (p *MemPostings) addFor(id storage.SeriesRef, l labels.Label) {
 }
 
 func (p *MemPostings) PostingsForLabelMatching(ctx context.Context, name string, match func(string) bool) Postings {
+	// We'll copy the values into a slice and then match over that,
+	// this way we don't need to hold the mutex while we're matching,
+	// which can be slow (seconds) if the match function is a huge regex.
+	// Holding this lock prevents new series from being added (slows down the write path)
+	// and blocks the compaction process.
+	//
+	// Also, benchmarking shows that first copying the values into a slice and then matching over that is
+	// faster than matching over the map keys directly, at least on AMD64.
```

Comment:
Side-note: I am suspicious of this - maybe it's an artefact of the branch prediction cache.

Reply:
The comment was there, I didn't validate it. I'll remove it since there's now a bigger reason for extracting the slice.

Reply:
Removed the comment.